hyperbolic-rs/preprocess.py

```diff
@@ -21,38 +21,114 @@ def movielens_to_dict(dataset_path):
+    :return: Dictionary containing users as keys, and a numpy array of items the user
+      interacted with, sorted by the time of interaction.
+    """
+    samples = {}
+    with open(file_path, 'r') as f:
+        next(f)
+        for line in f:
+            line = line.strip('\n').split(',')
+            uid = line[0]
+            iid = line[1]
+            if uid in samples:
+                samples[uid].add(iid)
+            else:
+                samples[uid] = {iid}
+
+    all_items = set()
+    for ints in samples.values(): all_items.update(ints)
+    interactions = sum([len(v) for v in samples.values()])
+    print(f"Processed users: {len(samples)}. Processed items: {len(all_items)}. Interactions: {interactions}")
+
+    filtered_samples = {uid: ints for uid, ints in samples.items() if len(ints) >= min_interactions}
+
+    filtered_items = set()
+    for ints in filtered_samples.values(): filtered_items.update(ints)
+    interactions = sum([len(v) for v in filtered_samples.values()])
+    print(f"After filtering: users: {len(filtered_samples)}, items: {len(filtered_items)}, interactions: {interactions}")
+
+    return filtered_samples
+
+
+def build_iid2title(item_id_key, item_title_key):
+    keen_metadata = "data/keen/keens_and_gems_25_june_2020.json"
+    iid2title = {}
+    with open(keen_metadata, "r") as f:
+        for line in f:
+            line = json.loads(line)
+            if item_id_key in line and item_title_key in line:
+                iid2title[line[item_id_key]] = line[item_title_key][1:-1]
+    return iid2title
+
+
+def map_item_ids_to_sequential_ids(samples):
+    """
+    :param samples: dict of <user_id>: [<item_id>, <item_id>, ...]
+    :return:
+    """
+    uid2id, iid2id = {}, {}
+    for uid, ints in samples.items():
+        if uid not in uid2id:
+            uid2id[uid] = len(uid2id)
+        for iid in ints:
+            if iid not in iid2id:
+                iid2id[iid] = len(iid2id)
+
+    return uid2id, iid2id
+
+
+def create_splits(samples, do_random=False):
     """
     Splits (user, item) dataset to train, dev and test.
 
     :param samples: Dict of sorted examples.
-    :param random: Bool whether to extract dev and test by random sampling. If False, dev, test are the last two
+    :param do_random: Bool whether to extract dev and test by random sampling. If False, dev, test are the last two
         items per user.
     :return: examples: Dictionary with 'train','dev','test' splits as numpy arrays
         containing corresponding (user_id, item_id) pairs, and 'to_skip' to a Dictionary containing filters
         for each user.
     """
     train, dev, test = [], [], []
     for uid, items in samples.items():
-        if random:
-            np.random.shuffle(items)
+        if do_random:
+            random.shuffle(items)
         test.append([uid, items[-1]])
         dev.append([uid, items[-2]])
         for iid in items[:-2]:
```

Comment:
What happened here?

Reply:
I changed the name of the argument so it doesn't clash with the Python module 'random'.
If random == True, then it shuffles the items. This will create a different train/dev/test split in every run. In my case, I always set it to False, so if you run this script twice, you will get the same splits.

Reply:
I think it might be better to provide a command line argument for the random seed. When it is None, you pick a random seed, and when it is provided you use it.

Reply:
Done